modern/org/commcare/modern/database/TableBuilder.java

```diff
@@ -0,0 +1,191 @@
+        rawCols.add(DatabaseHelper.ID_COL);
+
+        if(p instanceof IMetaData) {
+            String[] keys = ((IMetaData)p).getMetaDataFields();
+            for(String key : keys) {
+                String columnName = scrubName(key);
+                if(!rawCols.contains(columnName)) {
+                    rawCols.add(columnName);
+                    String columnDef = columnName;
+
+                    //Modifiers
+                    if (unique.contains(columnName)) {
+                        columnDef += " UNIQUE";
+                    }
+                    cols.add(columnDef);
+                }
+            }
+        }
+
+        cols.add(DatabaseHelper.DATA_COL + " BLOB");
+        rawCols.add(DatabaseHelper.DATA_COL);
+    }
+
+
+    HashSet<String> unique = new HashSet<String>();
+    public void setUnique(String columnName) {
+        unique.add(scrubName(columnName));
+    }
+
+    public String getTableCreateString() {
+
+        String built = "CREATE TABLE " + scrubName(name) + " (";
+        for(int i = 0 ; i < cols.size() ; ++i) {
+            built += cols.elementAt(i);
+            if(i < cols.size() - 1) {
+                built += ", ";
+            }
+        }
+        built += ");";
+        return built;
+    }
+
+    public Pair<String, List<Object>> getTableInsertData(Persistable p){
+
+        StringBuilder stringBuilder = new StringBuilder();
+
+        stringBuilder.append("INSERT INTO " + scrubName(name) + " (");
+        HashMap<String, Object> contentValues = DatabaseHelper.getMetaFieldsAndValues(p);
+
+        ArrayList<Object> params = new ArrayList<Object>();
+
+
+        for(int i = 0 ; i < rawCols.size() ; ++i) {
+            stringBuilder.append(rawCols.elementAt(i));
+            if(i < rawCols.size() - 1) {
+                stringBuilder.append(", ");
+            }
+        }
+
+        stringBuilder.append(") VALUES (");
+
+        for(int i = 0 ; i < rawCols.size() ; ++i) {
+            Object currentValue = contentValues.get(rawCols.elementAt(i));
+            stringBuilder.append("?");
+            params.add(currentValue);
+            if(i < rawCols.size() - 1) {
+                stringBuilder.append(", ");
+            }
+        }
+
+        stringBuilder.append(");");
+
+        return new Pair<String, List<Object>>(stringBuilder.toString(), params);
+    }
+
+    //sqlite doesn't like dashes
+    public static String scrubName(String input) {
+        return input.replace("-", "_");
+    }
+
+    public static byte[] toBlob(Persistable p){
+        ByteArrayOutputStream bos = new ByteArrayOutputStream();
+        OutputStream out = bos;
+        try {
+            p.writeExternal(new DataOutputStream(out));
+            out.close();
+        } catch (IOException e1) {
+            e1.printStackTrace();
+            throw new RuntimeException("Failed to serialize externalizable for content values");
+        }
+        return bos.toByteArray();
```

Comment:
I would remove `OutputStream out` and just use `bos` in its place. This might require you to create a `byte[]` variable to store `bos.toByteArray()` before you close `bos`, but I think maybe you should do that regardless because it feels weird to close the stream then read bytes out of it